### DISCUSSION TITLE
zc_install: Initial installation can result in database primed for "mixed collations"

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#5825] [Ideas]
Found this reviewing a zc156c initial installation.  The database was created using the `utf8_general_ci` collation:

![image](https://user-images.githubusercontent.com/2685585/73069185-13a98e00-3e7b-11ea-9a65-e4097998d38a.png)

During the initial install, the customer chose `utf8mb4` as the collation choice, resulting in tables using the aforementioned collation, e.g.

![image](https://user-images.githubusercontent.com/2685585/73069275-54a1a280-3e7b-11ea-9d62-2eef7783cbb2.png)

This has "primed" the database for future "mixed collations" issues, as any character-based fields added by plugins will result in those fields' collation being set to the database default (`utf8_general_ci`).

I'll suggest, at a minimum, that the `zc_install` process _warn_ of the situation or, alternately, let the installing human know the collation discovered in the database and continue (after confirmation) the installation using the database's base collation.

**Comment by torvista** [comment]
Newbie question: why can't the installer create the database, avoiding this?

**Comment by lat9** [comment]
My _feeling_ is that it's *complicated*, since the database's permissions, user and password need to be 'created' as well for the database to be accessible.

**Comment by drbyte** [comment]
> Newbie question: why can't the installer create the database, avoiding this?

Most managed servers don't allow database-creation privileges to their db users.

> My _feeling_ is that it's _complicated_, since the database's permissions, user and password need to be 'created' as well for the database to be accessible.

Yes, and that too.

**Comment by scottcwilson** [comment]
I like the idea of warning on detection of this condition.  It's almost certainly not what someone wants.

**Comment by lat9** [comment]
Actually, I _think_ it's just a matter (on that initial install) to, for example:
```
ALTER DATABASE `my_db_name` CHARACTER SET utf8mb4 COLLATION utf8mb4_general_ci
```

... or whatever character-set was chosen and the default-collation chosen.

**Comment by drbyte** [comment]
@lat9 I suspect you could add your logic into `class.zcDatabaseInstaller.php`, something like the following, after the `runZeroDateSql` function declaration:

```php
// approx line 98
    public function setDbCharsetAndCollation()
    {
        //read current collation
        $sql = 'SQL to read current collation/charset';
        $result = $this->db->execute($sql);
        if (!$result || $result->link->errno != 0) {
            $error = $this->db->error_number . ': ' . $this->db->error_text;
        }
        // parse $result here

        // generate new statement to set collation
        $sql = 'SQL STATEMENT HERE';
        $this->tryExecute($sql); // or call $this->>db->execute($sql) if you want to directly inspect the result
    }
 ```

I mentioned querying existing charset/collation because sometimes a db might already be set to a utf8 variant that is acceptable (thinking some Chinese db's that use more than `mb4` or other than `general_ci` which already handles even longer multibyte sets), in which case you might just skip making the change. Or perhaps the function could just be skipped if a DEV Mode constant is set somewhere, in case someone runs into a problem because of this.
Or maybe I'm overthinking.

And, you mentioned that you wanted to do this only on "install"; so probably best to call it in `AjaxLoadMainSql.php`:
```diff
$dbInstaller = new zcDatabaseInstaller($options);
$result = $dbInstaller->getConnection();
+$dbInstaller->setDbCharsetAndCollation();
$extendedOptions = array('doJsonProgressLogging'=>TRUE, 'doJsonProgressLoggingFileName'=>DEBUG_LOG_FOLDER . '/progress.json', 'id'=>'main', 'message'=>TEXT_CREATING_DATABASE);
$file = DIR_FS_INSTALL . 'sql/install/mysql_zencart.sql';
logDetails('processing file ' . $file);
```
